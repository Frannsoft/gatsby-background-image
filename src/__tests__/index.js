import "@babel/polyfill"
import { render, cleanup, fireEvent } from "react-testing-library"
import React from "react"
import BackgroundImage from "../"
import { createImageToLoad } from "../";
import getBackgroundStyles from "../BackgroundUtils";

afterAll(cleanup)

const fixedShapeMock = {
  width: 100,
  height: 100,
  src: `test_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  base64: `string_of_base64`,
}

const fluidShapeMock = {
  aspectRatio: 1.5,
  src: `test_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  sizes: `(max-width: 600px) 100vw, 600px`,
  base64: `string_of_base64`,
}

const setup = (fluid = false,
               addClass = false,
               additionalClass = ``,
               fixedClass = true,
               onLoad = () => {},
               onError = () => {}) => {

  if (addClass) {
    // Create the style class.
    const styleElement = document.createElement("style")
    styleElement.textContent = `
      .fixedImage {
        backgroundRepeat: 'repeat-y';
      }
    `
    document.body.appendChild(styleElement)
  }
  const classNames = fixedClass ?
      `fixedImage${additionalClass}` : additionalClass.trim()
  const { container } = render(
    <BackgroundImage
      backgroundColor
      className={classNames}
      style={{ display: `inline` }}
      title={`Title for the image`}
      alt={`Alt text for the image`}
      {...fluid && { fluid: fluidShapeMock }}
      {...!fluid && { fixed: fixedShapeMock }}
      onLoad={onLoad}
      onError={onError}
      placeholderStyle={{ color: `red` }}
      placeholderClassName={`placeholder`}
      classId="test"
    />
  )

  return container
}

describe(`<BackgroundImage />`, () => {
  it(`should render fixed size images`, () => {
    const component = setup()
    expect(component).toMatchSnapshot()
  })

  it(`should render fluid images`, () => {
    const component = setup(true)
    expect(component).toMatchSnapshot()
  })

  it(`should work with single external class`, () => {
    const component = setup(true)
    expect(component).toMatchSnapshot()
  })

  it(`should work with another external class`, () => {
    const component = setup(true, true, ` test`)
    expect(component).toMatchSnapshot()
  })

  it(`should create style tag with pseudo-elements`, () => {
    const styleTag = setup(true, true, ` test`).querySelector(`style`)
    expect(styleTag).toBeInTheDocument()
    expect(styleTag).toMatchSnapshot()
  })

  it(`should create style tag at fixed with pseudo-elements`, () => {
    const styleTag = setup(false, true, ` test`).querySelector(`style`)
    expect(styleTag).toBeInTheDocument()
    expect(styleTag).toMatchSnapshot()
  })

  it(`should have class with pseudo element in style tag`, () => {
    const styleTag = setup(true, true, ` test`).querySelector(`style`)
    expect(styleTag)
        .toHaveTextContent(`.gatsby-background-image-test:before`)
    expect(styleTag)
        .toHaveTextContent(`background-repeat: 'repeat-y';`)
  })

  it(`should parse background styles`, () => {
    const backgroundStyles = getBackgroundStyles(``)
    expect(backgroundStyles).toEqual({})
  })

  it(`should call onLoad and onError image events`, () => {
    const onLoadMock = jest.fn()
    const onErrorMock = jest.fn()
    const image = createImageToLoad({
      fluid: {
        src: ``,
      },
      onLoad: onLoadMock,
      onError: onErrorMock
    })
    fireEvent.load(image)
    fireEvent.error(image)

    expect(onLoadMock).toHaveBeenCalledTimes(1)
    expect(onErrorMock).toHaveBeenCalledTimes(1)
  })
})
