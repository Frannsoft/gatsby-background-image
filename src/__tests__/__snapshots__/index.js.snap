// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackgroundImage /> should create style tag at fixed with pseudo-elements 1`] = `
<style>
  
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                  content: "";
                  opacity: 0;
                }
              
</style>
`;

exports[`<BackgroundImage /> should create style tag with pseudo-elements 1`] = `
<style>
  
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                  content: "";
                  opacity: 0;
                }
              
</style>
`;

exports[`<BackgroundImage /> should render fixed size images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline;"
  >
    <style>
      
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                  content: "";
                  opacity: 0;
                }
              
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" /&gt;&lt;source srcSet="some srcSet" /&gt;&lt;img width="100" height="100" src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render fluid images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline;"
  >
    <style>
      
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                  content: "";
                  opacity: 0;
                }
              
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;source srcSet="some srcSet" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with another external class 1`] = `
<div>
  <div
    class="fixedImage test gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline;"
  >
    <style>
      
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: 'repeat-y';
                  content: "";
                  opacity: 0;
                }
              
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;source srcSet="some srcSet" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with single external class 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline;"
  >
    <style>
      
                .gatsby-background-image-test:before,
                .gatsby-background-image-test:after {
                  background-size: cover;
                  content: '';
                  display: block;
                  position: absolute;
                  width: 100%;
                  height: 100%;
                  top: 0;
                  left: 0;
                  transition: opacity 0.25s ease-in-out;
                  -webkit-transition: opacity 0.25s ease-in-out;
                  -moz-transition: opacity 0.25s ease-in-out;
                  -o-transition: opacity 0.25s ease-in-out;
                }
                .gatsby-background-image-test:before {
                  z-index: -101;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                }
                .gatsby-background-image-test:after {
                  z-index: -100;
                  background-image: url(test_image.jpg);
                  background-repeat: no-repeat;
                  content: "";
                  opacity: 0;
                }
              
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;source srcSet="some srcSet" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;
